core/virtualhostname/virtualhostname.go

```
@@ -33,6 +35,60 @@ const (
 	ContainerTarget
 )
 
+// NewInfoMachineTarget returns a new Info struct for a machine target.
+func NewInfoMachineTarget(modelUUID string, machine int) (Info, error) {
+	if !names.IsValidModel(modelUUID) {
+		return Info{}, errors.Errorf("invalid model UUID: %q", modelUUID)
+	}
+	if !names.IsValidMachine(fmt.Sprint(machine)) {
+		return Info{}, errors.Errorf("invalid machine number: %d", machine)
+	}
+	return newInfo(MachineTarget, modelUUID, machine, "", ""), nil
+}
+
+// NewInfoUnitTarget returns a new Info struct for a unit target.
+func NewInfoUnitTarget(modelUUID string, unit string) (Info, error) {
+	if !names.IsValidModel(modelUUID) {
+		return Info{}, errors.Errorf("invalid model UUID: %q", modelUUID)
+	}
+	if !names.IsValidUnit(unit) {
+		return Info{}, errors.Errorf("invalid unit name: %s", unit)
+	}
+	return newInfo(UnitTarget, modelUUID, 0, unit, ""), nil
+}
+
+// NewInfoContainerTarget returns a new Info struct for a container target.
+func NewInfoContainerTarget(modelUUID string, unit string, container string) (Info, error) {
+	if !names.IsValidModel(modelUUID) {
+		return Info{}, errors.Errorf("invalid model UUID: %q", modelUUID)
+	}
+	if !names.IsValidUnit(unit) {
+		return Info{}, errors.Errorf("invalid unit name: %s", unit)
+	}
+	return newInfo(ContainerTarget, modelUUID, 0, unit, container), nil
+}
+
+// newInfo returns a new Info struct for the given target.
+func newInfo(target HostnameTarget, modelUUID string, machine int, unit string, container string) Info {
+	info := Info{}
+	switch target {
+	case MachineTarget:
+		info.target = MachineTarget
+		info.modelUUID = modelUUID
+		info.machine = machine
+	case UnitTarget:
+		info.target = UnitTarget
+		info.modelUUID = modelUUID
+		info.unit = unit
+	case ContainerTarget:
+		info.target = ContainerTarget
+		info.modelUUID = modelUUID
+		info.unit = unit
+		info.container = container
+	}
+	return info
+}
```

Even though it's small, this should return an error, in case you get a target you don't expect. It'll prevent issues in the future.

---

I think because this function is not exported we don't need to return an error today if we don't need one. That can be changed in the future without affecting the package's api.

---

yes, this function is just used by the exported ones, so it shouldn't be called by a target i don't expect.

@SimonRichardson what's juju stance on `panic` in this case where we are sure that it shouldn't happen?

---

let's be defensive and return an error in a default case (when we get an unexpected target)